Bug 1910318: [OCPCLOUD-931] Add condition to show actuator exists output on machine status

JoelSpeed opened the pull request:
This adds conditions to the machine status, in particular it adds the `MachineExists` condition which should show whether or not the actuator exists function is working (else condition is unknown) and whether we consider the machine to exist or not yet.

JoelSpeed commented:
/retest

JoelSpeed commented:
I manually tested this and it's behaving as I would expect. I've also extended the testing to cover the lastUpdated field better so now we test more throughly when that field is updated

elmiko commented on pkg/apis/machine/v1beta1/condition_consts.go:
```diff
@@ -27,3 +27,18 @@ const (
+
+const (
+	// MachineExistsCondition is set on the Machine to show whether a virtual mahcine has been created by the cloud provider.
+	MachineExistsCondition ConditionType = "MachineExists"
```
given the language in these conditions around "machine" vs "instance", i would have expected this to be named `InstanceExistsCondition`. what's the reasoning behind calling this one "Machine" instead of "Instance" ?

JoelSpeed replied:
Instance seems like a good choice, I'll update this

alexander-demichev reviewed (approved):
thanks!

/approve

elmiko reviewed (commented):
the code generally looks good to me, i just have a question about the "machine" vs "instance" naming for one of the conditions.

lobziik commented on pkg/apis/machine/v1beta1/condition_consts.go:
```diff
@@ -27,3 +27,18 @@ const (
+
+const (
+	// MachineExistsCondition is set on the Machine to show whether a virtual mahcine has been created by the cloud provider.
+	MachineExistsCondition ConditionType = "MachineExists"
```
I'm bit concerned about mixing `Instance` and `Machine` there. Can we choose one word for this?

lobziik commented on pkg/controller/machine/controller.go:
```diff
@@ -421,61 +466,77 @@ func isInvalidMachineConfigurationError(err error) bool {
 }
 
-func (r *ReconcileMachine) setPhase(machine *machinev1.Machine, phase string, failureCause error) error {
+func (r *ReconcileMachine) updateStatus(machine *machinev1.Machine, phase string, failureCause error, originalConditions []machinev1.Condition) error {
```
WDYT about call `machine.GetConditions` inside this function and don't pass it here explicitly? As I see we return value from `GetConditions` method.

JoelSpeed replied:
The reason we pass it explicitly here is because we need to catch what the original state of the conditions is before we make any modifications to it. So getting the conditions once they've been passed into this function wouldn't work as we have already modified them on the machine by the time they get to this point.

I admit this isn't the easiest way this could be factored, but I think to make it simpler we would need to undertake quite a large refactor of the reconciliation logic to ensure we only update the status in one place - maybe a follow up for the not too distant future 🤔

elmiko replied:
maybe drop a todo comment in the code to keep track?

JoelSpeed commented:
@elmiko @lobziik I've updated the name as per your suggestion and added a comment to explain why we pass the conditions into updateStatus as we do today. I think the context of the TODO would be too long to add here so I think it's best to leave that out for now and raise an issue/jira instead

elmiko reviewed (approved):
thanks Joel!
/lgtm
